### ISSUE TITLE
Different AP after training

### ISSUE THREAD

**Issue by JulienDufour** [fizyr/keras-retinanet#549]
Hi,

I'm facing to a very strange behavior. I'm using the same evaluation method during the training and after. The results, in terms of average precision are different when I reload and evaluate my model. I'm using the same parameters and the same evaluation dataset during the evaluation after and during the training.

Training results : 
cabbage 0.9206
colza 0.8249
greensalad 0.9800
leekonion 0.7145
redsalad 0.9764
mAP: 0.8833
Epoch 00033: mAP improved from 0.88021 to 0.88327, saving model to ../snapshots/resnet101_csv.h5

Evaluation results (with ../snapshots/resnet101_csv.h5):
cabbage 0.7723
colza 0.6389
greensalad 0.9561
leekonion 0.3722
redsalad 0.9596
mAP: 0.7398

I tested two different ways to evaluate the model. But, both products the same result.
```
# create object that stores backbone information
backbone = models.backbone('resnet101')
backbone_retinanet = backbone.retinanet

# model = models.load_model(model_path, backbone_name='resnet101', convert=True)
model = backbone_retinanet(validation_generator.num_classes())
model.load_weights(model_path, by_name=True, skip_mismatch=True)

# make prediction model
from keras_retinanet.models.retinanet import retinanet_bbox
prediction_model = retinanet_bbox(model=model)

# Evaluate
average_precisions, recalls, precisions, infer_time = evaluate(validation_generator,
																prediction_model,
																score_threshold=0.3,
																max_detections=100, 
																iou_threshold=0.5)
```

**Comment by hgaiser** [comment]
I see you set the `score_threshold` to `0.3` in your evaluation. By default the `score_threshold` is 0.05, which is also the threshold used for evaluation while training. Could that be the cause of the difference?

**Comment by JulienDufour** [comment]
It is not the case because I used different parameters. I already check every parameters. Normally everything is similar. Below, a print of argparse.

```
Namespace(alpha=0.5, annotations='../dataset/annotation_train.csv', backbone='resnet101', batch_size=12, classes='../dataset/classes.csv', dataset_type='csv', epochs=500, evaluation=True, freeze_backbone=False, gamma=5.0, gpu=None, image_max_side=752, image_min_side=480, imagenet_weights=True, multi_gpu=0, multi_gpu_force=False, random_transform=True, score_threshold=0.3, sigma=3.0, snapshot='../snapshots/resnet101_csv.h5', snapshot_path='../snapshots', snapshots=True, steps=10000, tensorboard_dir='../logs', val_annotations='../dataset/annotation_val.csv', weights=None)
```

**Comment by JulienDufour** [comment]
You're right, but I had few parameters in order to control the evaluation and the focal loss from args.
```
    parser.add_argument('--gamma', help='Gamma (Focal Loss). RetinaNet is relatively robust to Gamma E [0.5,5].', type=float, dest='gamma', default=2.0)
    parser.add_argument('--alpha', help='Alpha (Focal Loss) E [0.1,0.999].', type=float, dest='alpha', default=0.25)
    parser.add_argument('--sigma', help='Sigma (Smooth L1 regression).', type=float, dest='sigma', default=3.0)
    parser.add_argument('--score-threshold', help='The score confidence threshold to use for detections. (Used during mAP evaluation)', type=float, dest='score_threshold', default=0.05)
```
I added "score_threshold" in order to get the same detections during training and inference.
```
 if args.evaluation and validation_generator:
        if args.dataset_type == 'coco':
            from ..callbacks.coco import CocoEval

            # use prediction model for evaluation
            evaluation = CocoEval(validation_generator, tensorboard=tensorboard_callback)
        else:
            evaluation = Evaluate(validation_generator, args.score_threshold, tensorboard=tensorboard_callback)
        evaluation = RedirectModel(evaluation, prediction_model)
        callbacks.append(evaluation)
```

**Comment by hgaiser** [comment]
There's your problem then, the second argument is the `iou_threshold`. It's better  to name the argument, like `score_threshold=score_threshold`. Then you know for sure you're setting the right thing.

**Comment by JulienDufour** [comment]
Thank you very much hgaiser ! I passed over this line several times... And, it was so obvious. I'm feeling worse than ever.

**Comment by JerryIndus** [comment]
> There's your problem then, the second argument is the iou_threshold. It's better  to name the argument, like score_threshold=score_threshold. Then you know for sure you're setting the right thing.

Excuse me, sorry to bother you. But I meet a problem which is similar with this one.  I separate my own dataset into training data, validation data and test data. And the mAP performs well on validation data. After i convert model into inference model, the mAP performs very bad on test data (even validation data ). I donnot know why.  When i reading this issue#549 , i guess, maybe i also need to get the same detections during training and inference. But i can't understand what do you say. Can you express it more explitly?
